tests/42tags.pl

```diff
@@ -0,0 +1,355 @@
+
+      my ( $user, $room_id );
+
+      matrix_register_user( $http, undef, with_events => 0 )->then( sub {
+         ( $user ) = @_;
+
+         matrix_create_room( $user );
+      })->then( sub {
+         ( $room_id ) = @_;
+
+         matrix_remove_tag( $user, $room_id, "test_tag" );
+      })->on_done( sub {
+         provide can_remove_tag => 1
+      });
+   };
+
+
+test "Can list tags for a room",
+   requires => [qw( first_api_client can_add_tag can_remove_tag )],
+
+   do => sub {
+      my ( $http ) = @_;
+
+      my ( $user, $room_id );
+
+      matrix_register_user( $http, undef, with_events => 0 )->then( sub {
+         ( $user ) = @_;
+
+         matrix_create_room( $user );
+      })->then( sub {
+         ( $room_id ) = @_;
+
+         matrix_add_tag( $user, $room_id, "test_tag", {} );
+      })->then( sub {
+         matrix_list_tags( $user, $room_id );
+      })->then( sub {
+         my ( $tags ) = @_;
+
+         keys %{ $tags } == 1 or die "Expected one tag for the room";
+         defined $tags->{test_tag} or die "Unexpected tag";
+
+         matrix_remove_tag( $user, $room_id, "test_tag" );
+      })->then( sub {
+         matrix_list_tags( $user, $room_id );
+      })->then( sub {
+         my ( $tags ) = @_;
+
+         keys %{ $tags } == 0 or die "Expected no tags for the room";
+
+         Future->done(1);
+      });
+   };
+
+
+test "Tags appear in the v1 /events stream",
+   requires => [qw( first_api_client can_add_tag can_remove_tag )],
+
+   do => sub {
+      my ( $http ) = @_;
+
+      my ( $user, $room_id );
+
+      matrix_register_user( $http, undef )->then( sub {
+         ( $user ) = @_;
+
+         matrix_create_room( $user );
+      })->then( sub {
+         ( $room_id ) = @_;
+
+         matrix_add_tag( $user, $room_id, "test_tag", { order => 1 } );
+      })->then( sub {
+         await_event_for( $user, sub {
+            my ( $event ) = @_;
+            return unless $event->{type} eq "m.tag"
+               and $event->{room_id} eq $room_id;
+
+            my %tags = %{ $event->{content}{tags} };
+            keys %tags == 1 or die "Expected exactly one tag";
+            defined $tags{test_tag} or die "Unexpected tag";
+            $tags{test_tag}{order} == 1 or die "Expected order == 1";
+
+            return 1;
+         });
+      });
+   };
+
+
+test "Tags appear in the v1 /initalSync",
+   requires => [qw( first_api_client can_add_tag can_remove_tag )],
+
+   do => sub {
+      my ( $http ) = @_;
+
+      my ( $user, $room_id );
+
+      matrix_register_user( $http, undef, with_events => 0 )->then( sub {
+         ( $user ) = @_;
+
+         matrix_create_room( $user );
+      })->then( sub {
+         ( $room_id ) = @_;
+
+         matrix_add_tag( $user, $room_id, "test_tag", { order => 1 } );
+      })->then( sub {
+         do_request_json_for( $user,
+            method => "GET",
+            uri    => "/api/v1/initialSync"
+        );
+      })->then( sub {
+         my ( $body ) = @_;
+
```

Comment:
Eh. I'd rather that we just logged the client requests on failure then rather than trying to cherry-pick which ones we thought it would be useful to log ahead of time.